sqe-http-api/Helpers/InputValidation.cs

```diff
@@ -0,0 +1,57 @@
+using System;
+using System.ComponentModel.DataAnnotations;
+using Newtonsoft.Json;
+using SQE.SqeHttpApi.DataAccess.Helpers;
+
+namespace SQE.SqeHttpApi.Server.Helpers
+{
+    public static class GeometryValidation
+    {
+        public static bool ValidateTransformMatrix(string transformMatrix)
+        {
+            try
+            {
+                _ = JsonConvert.DeserializeObject<transformMatrix>(transformMatrix);
+                return true;
+            }
+            catch
+            {
+                return false;
+            }
+        }
+
+        private class transformMatrix
+        {
+            [Required]
+            public double[][] matrix { get; set; }
+            
+            public transformMatrix(double[][] matrix)
+            {
+                if (!IsValidRows(matrix))
+                    throw StandardErrors.ImproperInputData("position");
+                if (!IsValidValues(matrix[0]) || !IsValidValues(matrix[1]))
+                    throw StandardErrors.ImproperInputData("position");
+            }
+            
+            bool IsValidRows(double[][] mat)
+            {
+                if (mat.Length != 2)
+                    return false;
+                if (mat[0].Length != 3)
+                    return false;
+                if (mat[1].Length != 3)
+                    return false;
+                return true;
+            }
+
+            bool IsValidValues(double[] row)
+            {
+                if (row[0] < -1 && row[0] > 1)
+                    return false;
+                if (row[1] < -1 && row[1] > 1)
+                    return false;
+                return (int)row[2] == row[2];
```

Comment:
Are you testing that `row[2]` contains an integer? This may fail due to rounding errors. It will also fail with negative numbers - (int)-0.99999 is 0 and not -1. 

Better check that Math.Abs(Math.round(row[2]) - row[2]) < 1e-6`  (I'm using Round so it works with negative numbers as well)

Reply:
Also, note that I'm asking what it is you're testing - that's because a comment would have helped me.

Reply:
The test here is to make sure that the number in row[2] is a whole number.  So if row[2] is -0.999 (which is an invalid value), then (int)row[2] == row[2] -> 0 == -0.999 -> false, which is indeed the behavior I want (some documentation would indeed have helped there, sorry).  Another example where row[2] is -1.000, then (int)row[2] == row[2] -> -1 == -1.000 -> true. Basically, the test ensures that there are no significant decimal values, since JSON only has Number type, not int/float, and I cannot otherwise ensure that this value is the equivalent of an int (even if it has trailing 0's).

Reply:
JSON has a number type, but C# has ints. If you put an `int[][]` in the DTO, aren't floats going to be rejected by ASP.NET Core?

Reply:
I think it isn't a real problem, the transform matrix in JSON is a 2d array of Numbers, the DTO just uses receives that JSON string and inserts it as-is into the database, where this data is also just a string, so the data is never converted into a C# data structure for insertion into the database.  
The point of the validator here is to make sure that any submitted transform matrix string is in fact 1. valid json, 2. having a "matrix" property that is a 2d array of doubles, 3. which has exactly two first level members of the array (i.e., [[],[]] and not [[],[],[]], etc.), 4. each of which has exactly three members, 5. the first two of which fall within a specific a specific range of values (I need to recheck what is a valid range for these), and 6. the third member is a whole number—it doesn't matter if it is 3.000 or 3, I just want it to be a whole number because sub-pixel translation makes no sense here (e.g., translate on the X axis by 5493.8796 pixels).
I will add all this to the function to clarify and also I will add it to the API documentation.

Reply:
OK.